File: crates/ci-python/test/test_golden.py
"""Golden parity test for the data-bound Python bindings.

Loads the shared cross-language fixture ``tests/fixtures/golden.json`` and, for
each of the seven tests, builds a :class:`Dataset` from the case's columns,
constructs the test with the case's parameters, runs it, and asserts that
``statistic`` / ``p_value`` / ``dof`` match the recorded ``expected`` values.
This is the binding's numeric parity gate against the scipy/pgmpy reference.
"""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any

import numpy as np
import pytest

from ci_python import (
    ChiSquared,
    CressieRead,
    Dataset,
    FreemanTukey,
    LogLikelihood,
    ModifiedLikelihood,
    PearsonCorrelation,
    PearsonEquivalence,
)

# Locate the repo-root fixture relative to this file:
# crates/ci-python/test/test_golden.py -> repo root is three parents up.
REPO_ROOT = Path(__file__).resolve().parents[3]
GOLDEN_PATH = REPO_ROOT / "tests" / "fixtures" / "golden.json"

TOL = 1e-7
EXPECTED_CASE_COUNT = 73

# Map the fixture's stable test name to its binding class.
TEST_CLASSES = {
    "chi_squared": ChiSquared,
    "log_likelihood": LogLikelihood,
    "cressie_read": CressieRead,
    "freeman_tukey": FreemanTukey,
    "modified_likelihood": ModifiedLikelihood,
    "pearson_correlation": PearsonCorrelation,
    "pearson_equivalence": PearsonEquivalence,
}


def _load_cases() -> list[dict[str, Any]]:
    with GOLDEN_PATH.open() as fh:
        return json.load(fh)


def _build_dataset(columns: dict[str, dict[str, Any]]) -> Dataset:
    spec = {
        name: (col["kind"], np.asarray(col["values"], dtype=np.float64))
        for name, col in columns.items()
    }
    return Dataset(spec)


def _construct(name: str, data: Dataset, params: dict[str, Any]) -> Any:
    cls = TEST_CLASSES[name]
    if name == "pearson_correlation":
        return cls(data)
    if name == "pearson_equivalence":
        return cls(data, delta_threshold=params["delta_threshold"])
    # Discrete power-divergence family: configured by `yates`.
    return cls(data, yates=params["yates"])


def _assert_close(actual: float | None, expected: Any, field: str, case_id: str) -> None:
    if expected is None:
        # Null fields are skipped (the test does not define them).
        return
    assert actual is not None, f"{case_id}: expected {field}={expected}, got None"
    exp = float(expected)
    if math.isinf(exp) or math.isnan(exp):
        # Defensive handling for inf statistic / nan (none in the current fixture).
        assert math.isinf(actual) == math.isinf(exp), f"{case_id}: {field} inf mismatch"
        assert math.isnan(actual) == math.isnan(exp), f"{case_id}: {field} nan mismatch"
        if math.isinf(exp):
            assert math.copysign(1.0, actual) == math.copysign(1.0, exp), (
                f"{case_id}: {field} inf sign mismatch"
            )
        return
    assert actual == pytest.approx(exp, abs=TOL, rel=0.0), (
        f"{case_id}: {field} mismatch: got {actual!r}, expected {exp!r}"
    )


GOLDEN_CASES = _load_cases()


@pytest.mark.parametrize(
    ("index", "case"),
    list(enumerate(GOLDEN_CASES)),
    ids=[f"{i:02d}-{c['test']}" for i, c in enumerate(GOLDEN_CASES)],
)
def test_golden_case(index: int, case: dict[str, Any]) -> None:
    """Each fixture case reproduces its recorded statistic / p_value / dof."""
    case_id = f"case[{index}] {case['test']}"
    data = _build_dataset(case["columns"])
    test = _construct(case["test"], data, case["params"])

    result = test.run_test(case["x"], case["y"], case["z"])
    expected = case["expected"]

    _assert_close(result.statistic, expected.get("statistic"), "statistic", case_id)

    # p-value: handle exact-zero / inf defensively, otherwise compare within TOL.
    exp_p = float(expected["p_value"])
    if exp_p == 0.0:
        assert result.p_value == pytest.approx(0.0, abs=TOL), (
            f"{case_id}: p_value expected ~0, got {result.p_value!r}"
        )
    else:
        _assert_close(result.p_value, exp_p, "p_value", case_id)

    _assert_close(
        None if result.dof is None else float(result.dof),
        expected.get("dof"),
        "dof",
        case_id,
    )


def test_golden_covers_all_cases() -> None:
    """The fixture has exactly the expected number of cases, all dispatched."""
    assert len(GOLDEN_CASES) == EXPECTED_CASE_COUNT, (
        f"expected {EXPECTED_CASE_COUNT} cases, found {len(GOLDEN_CASES)}"
    )
    names = {c["test"] for c in GOLDEN_CASES}
    assert names == set(TEST_CLASSES), (
        f"fixture tests {names} do not match binding classes {set(TEST_CLASSES)}"
    )

File: crates/citest-r/tests/testthat/test-golden.R
# Golden parity test for the data-bound R bindings.
#
# Loads the packaged copy of the shared cross-language golden fixture and, for
# each of the eight tests, builds a `dataset()` from the case's columns,
# constructs the test with the case's parameters, runs it, and asserts that
# `statistic` / `p_value` / `dof` / `effect_size` match the recorded `expected`
# values within a tight tolerance. This is the binding's numeric parity gate
# against the scipy/pgmpy reference.

library(citest)

tol <- 1e-7
expected_case_count <- 80L

# Map the fixture's stable test name to its binding factory function.
test_factories <- list(
  chi_squared = chi_squared,
  log_likelihood = log_likelihood,
  cressie_read = cressie_read,
  freeman_tukey = freeman_tukey,
  modified_likelihood = modified_likelihood,
  pearson_correlation = pearson_correlation,
  fisher_z = fisher_z,
  pearson_equivalence = pearson_equivalence
)

find_golden <- function() {
  testthat::test_path("fixtures", "golden.json")
}

load_cases <- function() {
  jsonlite::fromJSON(find_golden(), simplifyVector = FALSE)
}

# Build a `citest_dataset` from a fixture case's `columns` map. Discrete columns
# are coerced to integer (so `dataset()` infers "discrete"); continuous columns
# stay double ("continuous").
build_dataset <- function(columns) {
  cols <- lapply(columns, function(col) {
    values <- as.numeric(unlist(col$values))
    if (identical(col$kind, "discrete")) {
      as.integer(values)
    } else {
      as.double(values)
    }
  })
  df <- as.data.frame(cols, stringsAsFactors = FALSE, check.names = FALSE)
  dataset(df)
}

construct_test <- function(name, data, params) {
  factory <- test_factories[[name]]
  if (name %in% c("pearson_correlation", "fisher_z")) {
    factory(data)
  } else if (name == "pearson_equivalence") {
    factory(data, delta_threshold = params$delta_threshold)
  } else {
    # Discrete power-divergence family: configured by `yates`.
    factory(data, yates = params$yates)
  }
}

# Compare one numeric field against its expected value. Null expectations
# require a NULL result; Inf / NaN are matched defensively.
assert_close <- function(actual, expected, field, case_id) {
  if (is.null(expected)) {
    expect_null(actual,
      label = sprintf("%s: expected %s=NULL, got %s", case_id, field, actual)
    )
    return(invisible())
  }
  expect_false(is.null(actual),
    label = sprintf("%s: expected %s=%s, got NULL", case_id, field, expected)
  )
  exp <- as.numeric(expected)
  act <- as.numeric(actual)
  if (is.infinite(exp) || is.nan(exp)) {
    expect_equal(is.infinite(act), is.infinite(exp),
      label = sprintf("%s: %s inf mismatch", case_id, field)
    )
    expect_equal(is.nan(act), is.nan(exp),
      label = sprintf("%s: %s nan mismatch", case_id, field)
    )
    if (is.infinite(exp)) {
      expect_equal(sign(act), sign(exp),
        label = sprintf("%s: %s inf sign mismatch", case_id, field)
      )
    }
    return(invisible())
  }
  expect_equal(act, exp,
    tolerance = tol,
    label = sprintf("%s: %s got %s, expected %s", case_id, field, act, exp)
  )
}

golden_cases <- load_cases()

test_that("golden fixture covers exactly the expected cases", {
  expect_equal(length(golden_cases), expected_case_count)
  names_seen <- unique(vapply(golden_cases, function(c) c$test, character(1)))
  expect_setequal(names_seen, names(test_factories))
})

test_that("each golden case reproduces its complete recorded result", {
  for (i in seq_along(golden_cases)) {
    case <- golden_cases[[i]]
    case_id <- case$id
    expect_true(nzchar(case_id), label = "fixture case ID must not be empty")
    expect_true(case$test %in% names(test_factories),
      label = sprintf("%s: unknown test %s", case_id, case$test)
    )

    data <- build_dataset(case$columns)
    test <- construct_test(case$test, data, case$params)

    z <- as.character(unlist(case$z))
    if (length(z) == 0L) {
      z <- character()
    }
    result <- run_test(test, case$x, case$y, z)
    expected <- case$expected

    assert_close(result$statistic, expected$statistic, "statistic", case_id)

    assert_close(result$p_value, expected$p_value, "p_value", case_id)

    dof_actual <- if (is.null(result$dof)) NULL else as.numeric(result$dof)
    assert_close(dof_actual, expected$dof, "dof", case_id)
    assert_close(
      result$effect_size,
      expected$effect_size,
      "effect_size",
      case_id
    )
  }
})
